src/ExpressionEvaluator/Core/Source/ResultProvider/Helpers/TypeHelpers.cs

```diff
@@ -626,24 +667,39 @@ internal static DkmCustomUIVisualizerInfo[] GetDebuggerCustomUIVisualizerInfo(th
                         continue;
                     }
 
-                    builder.Add(DkmCustomUIVisualizerInfo.Create((uint)builder.Count,
-                        visualizerAttribute.VisualizerDescription,
-                        visualizerAttribute.VisualizerDescription,
-                        // ClrCustomVisualizerVSHost is a registry entry that specifies the CLSID of the
-                        // IDebugCustomViewer class that will be instantiated to display the custom visualizer.
-                        "ClrCustomVisualizerVSHost",
-                        visualizerAttribute.UISideVisualizerTypeName,
-                        visualizerAttribute.UISideVisualizerAssemblyName,
-                        visualizerAttribute.UISideVisualizerAssemblyLocation,
-                        visualizerAttribute.DebuggeeSideVisualizerTypeName,
-                        visualizerAttribute.DebuggeeSideVisualizerAssemblyName));
+                    // We need to watch out for duplicates since derived objects might return the same
+                    // attributes as their parents.
+                    if (!visualizerAttributeBuilder.Contains(visualizerAttribute))
+                    {
+                        visualizerAttributeBuilder.Add(visualizerAttribute);
+                    }
                 }
 
                 underlyingType = underlyingType.GetBaseTypeOrNull(appDomain, out type);
             }
 
-            var result = (builder.Count > 0) ? builder.ToArray() : null;
-            builder.Free();
+            var visualizerInfoBuilder = ArrayBuilder<DkmCustomUIVisualizerInfo>.GetInstance();
+            foreach (var visualizerAttribute in visualizerAttributeBuilder)
```

Comment:
>visualizerAttributeBuilder [](http://example.com/codeflow?start=48&length=26)

We're relying on the `HashSet<T>` enumerator which may not return the attributes in the order they were added. We should consider having the `while` loop above use an `ArrayBuilder<T>` for collecting the attributes (as before), but also track the attributes in a `HashSet<T>` to avoid duplicates.

```csharp
var builder = ArrayBuilder<DkmCustomUIVisualizerInfo>.GetInstance();
HashSet<DkmClrDebuggerVisualizerAttribute> seen = null;
...
while (...)
{
    foreach (var attribute in type.GetEvalAttributes())
    {
        ...
        var visualizerAttribute = ...;
        if (seen is null)
        {
            seen = new HashSet<DkmClrDebuggerVisualizerAttribute>(DkmClrDebuggerVisualizerAttributeComparer.Instance);
        }
        else if (seen.Contains(visualizerAttribute))
        {
            continue;
        }
        builder.Add(DkmCustomUIVisualizerInfo.Create(...));
    }
}
```